libshpool/src/daemon/pager.rs

```diff
@@ -0,0 +1,371 @@
+
+use crate::{consts, protocol, tty};
+
+// poll relatively quickly to pick up pager exits reasonably fast,
+// but still slow enough to spend most of the time parked.
+const POLL_MS: i32 = 100;
+
+// Handles for poking at the pager from separate connection handlers.
+// This is basically the same idea as for ReaderCtl.
+#[derive(Debug)]
+pub struct PagerCtl {
+    /// Used to signal size changes so we can correctly trigger
+    /// a SIGWINCH on the pty.
+    pub tty_size_change: crossbeam_channel::Sender<tty::Size>,
+    /// Acks the completion of a resize/SIGWINCH.
+    pub tty_size_change_ack: crossbeam_channel::Receiver<()>,
+}
+
+/// Pager is capable of displaying a message to the user via
+/// the pager of their choice. Display will block until they
+/// quick or the connection drops.
+pub struct Pager {
+    /// The name of the pager program to use. Typically "less".
+    pager_bin: String,
+}
+
+impl Pager {
+    /// Create a new Pager.
+    pub fn new(pager_bin: String) -> Self {
+        Pager { pager_bin }
+    }
+
+    /// Display the message, blocking until the user quits
+    /// out of the pager process or the connection drops.
+    ///
+    /// Though this returns any anyhow::Result, you likely want to
+    /// downcast any errors to check to see if you got a PagerError,
+    /// since some error conditions ought to be handled by the caller.
+    /// In particular, you can't assume that the client_stream is still
+    /// healthy after a call to display() and you should check for
+    /// PagerError::ClientHangup to determine if you should continue
+    /// on with the connection.
+    ///
+    /// On success, returns the final tty size, since the user may
+    /// have resized their terminal while looking at the pager.
+    pub fn display(
+        &self,
+        // The client connection on which to display the pager.
+        client_stream: &mut UnixStream,
+        // The slot to install the control handle in
+        ctl_slot: Arc<Mutex<Option<PagerCtl>>>,
+        // The size of the tty to start off with
+        init_tty_size: tty::Size,
+        // The message to display
+        msg: &str,
+    ) -> anyhow::Result<tty::Size> {
+        let (tty_size_change_tx, tty_size_change_rx) = crossbeam_channel::bounded(0);
+        let (tty_size_change_ack_tx, tty_size_change_ack_rx) = crossbeam_channel::bounded(0);
+        {
+            let mut ctl_handle = ctl_slot.lock().unwrap();
+            if ctl_handle.is_some() {
+                return Err(anyhow!("only one pager per session at a time allowed"));
+            }
+
+            *ctl_handle = Some(PagerCtl {
+                tty_size_change: tty_size_change_tx,
+                tty_size_change_ack: tty_size_change_ack_rx,
+            });
+        }
+        // make sure we reset the control handles
+        let _ctl_guard = PagerCltGuard { ctl_slot };
+
+        let mut msg_file = tempfile::NamedTempFile::with_prefix("shpool_pager")
```

Comment:
Why not use a pipe to stdin?

Reply:
A pipe to stdin would be nice, but it is somewhat annoying to arrange. I can't just use '.stdin(Stdio::piped())` because I'm not spawning the command in the usual way and am instead execing it inside the pty fork. The next most obvious thing is to just write the data to the stdin of the child process, but the trouble is that we only have a handle to the pty wrapping the child process, not directly to the less process. I tried working around this with a manual pipe() call and using dup2() to try to attach the pipe right before execing, but I couldn't get that to work (I did bump the nix version in order to try this, which I want to keep in now that I've done it since it was a fair amount of work).